utils/spdx/src/main/kotlin/model/SpdxPackage.kt

```diff
@@ -192,6 +192,10 @@ data class SpdxPackage(
             "The SPDX ID '$spdxId' has to start with '${SpdxConstants.REF_PREFIX}'."
         }
 
+        require(spdxId.all { it.isLetterOrDigit() || it == '.' || it == '-' }) {
```

Comment:
I do recall a discussion from long time ago. It was about what kind of checks to do in which of the following ways:

1. Check failing deserialization
2. A separate verifier / linter function which highlights issues found the the document.

IIRC there was an agreement to not implement all strictness as type 1 check, but do some checks as type 2 check. I believe this check should fall under type 2. What do you think?

Reply:
While I agree that it makes sense to have such a separation, I'd say that all syntax checks (like invalid chars) should be done as 1, and all semantic checks (like a relation to a package that does not exist) should be done as 2. As such, I'd prefer to *have* this check, esp. as we already do other syntax checks on the SPDX ID, like requiring it to start with `SPDXRef-`.

Reply:
Have you considered that each time one chooses 1, it leads to failed de-serialization, which in turn leads to an analyzer error in case of the SPDX analyzer - While it would be possible to tolerate this kind of problem and analyze without an `issue`?

Reply:
Yes, I'm aware of that. But the same holds true for the "SPDXRef-" prefix: Theoretically we should live without it, but we still require it as the SPDX specs also do. So where do we draw the line?

IMO it's not ORT's responsibility to be lenient WRT such syntax errors. On the contrary, if ORT was lenient, probably more related tools will be lenient, too, forcing more and more tools to also become lax in this regard, ultimately resulting in many invalid SPDX files out there in the wild.

So, in a similar way like the analyzer in general relies on and implicitly enforces to follow best engineering practices by properly configuring your build system for example, we should also help the SPDX community by enforcing correct SPDX syntax here.

Reply:
> forcing more and more tools to also become lax

I believe you misunderstood what I was trying to say. My point was solely about tolerating such issues when de-serializing stuff. Not writing out SPDX file with such problems. Being able to de-serialize files with such issues does not force any tool to become "lax"

> we should also help the SPDX community by enforcing correct SPDX syntax here

That's possible in both ways, 1 and 2. While 1 is a tad less flexible.

> So, in a similar way like the analyzer in general relies on and implicitly enforces to follow best engineering practices by properly configuring your build system for example, we should also help the SPDX community by enforcing correct SPDX syntax here.

Well, when we're talking about just this ID case I'm ok. But I really fear the checks we built in which fail deserialization will become more and more strict, which in turn can waste time of compliance engineers using ORT working on issues which are not even necessary to fix and could be tolerated. Personally, I find referential integrity more important than the syntax within an ID. Anyhow, let's just stick to as it is now, but let's not take it as agreement to put in place any kind of more strict checks in the future to fail de-serialization.

Reply:
> Not writing out SPDX file with such problems. Being able to de-serialize files with such issues does not force any tool to become "lax"

Correct, that was an error in my thought.

> But I really fear the checks we built in which fail deserialization will become more and more strict

We absolutely should have as strict checks as possible, IMO, to discover errors early in a processing pipeline.

> working on issues which are not even necessary to fix

The question remains: Where do we draw the line then? Shouldn't we then also allow SPDX IDs *not* starting with "SPDXRef-"? And simply because there IMO is no clear answer to this question, we should stick to the safe side and adhere to the specs.

> Anyhow, let's just stick to as it is now, but let's not take it as agreement to put in place any kind of more strict checks in the future to fail de-serialization.

So, does that mean you approve this PR or not?